src/p2p/swarm.rs

```diff
@@ -1,22 +1,104 @@
 use crate::subscription::{SubscriptionFuture, SubscriptionRegistry};
+use anyhow::anyhow;
 use core::task::{Context, Poll};
 use libp2p::core::{
     connection::ConnectionId, multiaddr::Protocol, ConnectedPoint, Multiaddr, PeerId,
 };
 use libp2p::swarm::protocols_handler::{
     DummyProtocolsHandler, IntoProtocolsHandler, ProtocolsHandler,
 };
-use libp2p::swarm::{self, DialPeerCondition, NetworkBehaviour, PollParameters, Swarm};
+use libp2p::swarm::{self, NetworkBehaviour, PollParameters, Swarm};
 use std::collections::{HashMap, HashSet, VecDeque};
+use std::convert::TryFrom;
 use std::time::Duration;
+use std::{fmt, str::FromStr};
+
+/// A wrapper for `Multiaddr` that does **not** contain `Protocol::P2p`.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct MultiaddrWoPeerId(Multiaddr);
+
+impl From<Multiaddr> for MultiaddrWoPeerId {
+    fn from(addr: Multiaddr) -> Self {
+        Self(
+            addr.into_iter()
+                .filter(|p| !matches!(p, Protocol::P2p(_)))
+                .collect(),
```

Comment:
This will not work for those p2p-circuit multiaddrs correctly at all. It should be a single pop though I dislike this lenient func either way.

Reply:
Meh, it could also filter out `Protocol::P2pCircuit`; as for liking this or not this is just a matter of opinion, as one _can_ infallibly convert a `Multiaddr` containing `P2p` protocols into one that doesn't (as long as there are other protocols like `Tcp`); making the `Multiaddr` > `MultiaddrWithoutPeerId` fallible makes me wonder about error-handling in the `NetworkBehaviour::inject_*` functions - for now we can probably safely assume they don't contain the `P2p` protocols, since the `PeerId` is passed separately?